inst/include/dust/random/utils.hpp

```diff
@@ -0,0 +1,74 @@
+#ifndef DUST_RANDOM_UTILS_HPP
+#define DUST_RANDOM_UTILS_HPP
+
+#include <dust/random/numeric.hpp>
+
+namespace dust {
+namespace random {
+
+enum xoshiro_mode {STARSTAR, PLUSPLUS, PLUS};
+
+template <typename T, typename U>
+T int_to_real(U value);
+
+template <>
+inline HOSTDEVICE
+double int_to_real(uint64_t value) {
+  return double(value) / double(utils::uint64_max());
+}
+
+template <>
+inline HOSTDEVICE
+float int_to_real(uint64_t value) {
+  return float(value) / float(utils::uint64_max());
+}
+
+template <>
+inline HOSTDEVICE
+float int_to_real(uint32_t value) {
+  return float(value) / float(utils::uint32_max());
+}
+
+inline HOSTDEVICE
+uint64_t rotl(const uint64_t x, int k) {
+  return (x << k) | (x >> (64 - k));
+}
+
+inline HOSTDEVICE
+uint32_t rotl(const uint32_t x, int k) {
+  return (x << k) | (x >> (32 - k));
+}
+
+// This is probably doable with sizeof(T) * 8 too?
+template <typename T>
+constexpr size_t bit_size();
```

Comment:
Similar here, is this needed?

Reply:
yes, same reason above. With C++14 we could do templated constants which would be nicer but seems a weak reason to change standard